activation/nipost_test.go

```diff
@@ -807,6 +807,344 @@ func TestNIPSTBuilder_PoetUnstable(t *testing.T) {
+		)
+		mclock := defaultLayerClockMock(ctrl)
+
+		postClient := NewMockPostClient(ctrl)
+		nonce := types.VRFPostIndex(1)
+		postClient.EXPECT().Proof(gomock.Any(), gomock.Any()).Return(
+			&types.Post{
+				Indices: []byte{1, 2, 3},
+			}, &types.PostInfo{
+				Nonce: &nonce,
+			}, nil,
+		)
+
+		postService := NewMockpostService(ctrl)
+		postService.EXPECT().Client(sig.NodeID()).Return(postClient, nil).AnyTimes()
+
+		// successfully registered to 1 poet
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(10 * time.Second),
+		})
+		require.NoError(t, err)
+
+		// successful post exec
+		nb, err := NewNIPostBuilder(
+			db,
+			postService,
+			zaptest.NewLogger(t),
+			PoetConfig{},
+			mclock,
+			nil,
+			WithPoetServices(poetProver, addedPoetProver), // add both poet provers
+		)
+		require.NoError(t, err)
+
+		nipost, err := nb.BuildNIPost(context.Background(), sig, challengeHash, &types.NIPostChallenge{PublishEpoch: 7})
+		require.NoError(t, err)
+		require.NotNil(t, nipost)
+	})
+
+	t.Run("completely changed poet service BEFORE round started -> register new poet", func(t *testing.T) {
+		db := localsql.InMemory()
+
+		require.NoError(t, nipost.AddChallenge(db, sig.NodeID(), &challenge))
+		challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+		ctrl := gomock.NewController(t)
+		addedPoetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr2)
+		addedPoetProver.EXPECT().Proof(gomock.Any(), "").AnyTimes().Return(
+			&types.PoetProof{}, []types.Hash32{
+				challengeHash,
+				types.RandomHash(),
+				types.RandomHash(),
+			}, nil,
+		)
+		mclock := defaultLayerClockMock(ctrl)
+
+		postClient := NewMockPostClient(ctrl)
+		nonce := types.VRFPostIndex(1)
+		postClient.EXPECT().Proof(gomock.Any(), gomock.Any()).Return(
+			&types.Post{
+				Indices: []byte{1, 2, 3},
+			}, &types.PostInfo{
+				Nonce: &nonce,
+			}, nil,
+		)
+
+		postService := NewMockpostService(ctrl)
+		postService.EXPECT().Client(sig.NodeID()).Return(postClient, nil).AnyTimes()
+
+		// successfully registered to removed poet
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(10 * time.Second),
+		})
+		require.NoError(t, err)
+
+		// successful post exec
+		nb, err := NewNIPostBuilder(
+			db,
+			postService,
+			zaptest.NewLogger(t),
+			PoetConfig{},
+			mclock,
+			nil,
+			WithPoetServices(addedPoetProver), // add new poet
+		)
+		require.NoError(t, err)
+
+		nipost, err := nb.BuildNIPost(context.Background(), sig, challengeHash, &types.NIPostChallenge{PublishEpoch: 7})
+		require.NoError(t, err)
+		require.NotNil(t, nipost)
+	})
+
+	t.Run("1 poet added AFTER round started -> too late to register to added poet, get proof from known",
```

Comment:
I think this test doesn't match the description. It adds 2 registrations but later passes only 1 poet to the builder.

Reply:
right, my fault